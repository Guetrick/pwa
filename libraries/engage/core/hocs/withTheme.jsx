import React from 'react';
import { ThemeContext } from '@shopgate/pwa-common/context';

/**
 * Gets the context props for the wrapped component.
 * @param {Object} context The context.
 * @param {string|null} prop The optional prop name.
 * @returns {Object}
 */
const getInjectedProps = (context, prop) => {
  if (!prop) {
    return context;
  }

  return {
    [prop]: { ...context },
  };
};

/**
 * Injects the theme API into the desired component. This does not include the contexts.
 * @param {Function} WrappedComponent The react component to wrap.
 * @param {Object} [options={}] Options for the HOC.
 * @param {string} [options.prop] An optional prop name to inject the theme properties.
 * @returns {JSX}
 */
export function withTheme(WrappedComponent, options = {}) {
  /**
   * The actual HOC.
   * @param {Object} props The component props.
   * @returns {JSX}
   */
  const WithTheme = props => (
    <ThemeContext.Consumer>
      {({ contexts, ...themeContext }) => ( // The contexts are left out in favor of other HOCs.
        <WrappedComponent {...getInjectedProps(themeContext, options.prop)} {...props} />
      )}
    </ThemeContext.Consumer>
  );

  const displayName = WrappedComponent.displayName || WrappedComponent.name || 'Component';
  WithTheme.displayName = `WithTheme(${displayName})`;

  return WithTheme;
}
