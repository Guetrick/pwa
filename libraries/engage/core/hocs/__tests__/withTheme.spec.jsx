import React from 'react';
import { mount } from 'enzyme';
import { withTheme } from '../withTheme';

const mockTheme = {
  ComponentOne: function ComponentOne() { return (<div />); },
  ComponentTwo: function ComponentTwo() { return (<div />); },
};

const mockContext = jest.fn().mockReturnValue(mockTheme);

jest.mock('@shopgate/pwa-common/context', () => ({
  ThemeContext: {
    Consumer: ({ children }) => children(mockContext()),
  },
}));

// eslint-disable-next-line require-jsdoc
const MockComponent = () => null;

describe('engage > core > hocs > withTheme', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('should inject the context properties into the component', () => {
    const ComposedComponent = withTheme(MockComponent);
    const wrapper = mount(<ComposedComponent someProp />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).props()).toEqual({
      someProp: true,
      ...mockTheme,
    });
  });

  it('should inject a single property with the context into the component', () => {
    const ComposedComponent = withTheme(MockComponent, { prop: 'theme' });
    const wrapper = mount(<ComposedComponent someProp />);
    expect(wrapper).toMatchSnapshot();
    expect(wrapper.find(MockComponent).props()).toEqual({
      someProp: true,
      theme: {
        ...mockTheme,
      },
    });
  });
});
