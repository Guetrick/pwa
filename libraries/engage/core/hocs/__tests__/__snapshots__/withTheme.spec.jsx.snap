// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`engage > core > hocs > withTheme should inject a single property with the context into the component 1`] = `
<WithTheme(MockComponent)
  someProp={true}
>
  <Consumer>
    <MockComponent
      someProp={true}
      theme={
        Object {
          "ComponentOne": [Function],
          "ComponentTwo": [Function],
        }
      }
    />
  </Consumer>
</WithTheme(MockComponent)>
`;

exports[`engage > core > hocs > withTheme should inject the context properties into the component 1`] = `
<WithTheme(MockComponent)
  someProp={true}
>
  <Consumer>
    <MockComponent
      ComponentOne={[Function]}
      ComponentTwo={[Function]}
      someProp={true}
    />
  </Consumer>
</WithTheme(MockComponent)>
`;
